File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第9章编程作业.py
#编程作业 lanqiaoOJ 1433
N=1000006
Next = [0]*N
slen,plen=0,0
s = [0]*N
p = [0]*N
pre = [0]*29
def check(a,b,j):
    if a<0 or b<0: return a==b
    return a==b or (b==0 and a>j)
def inits():
    global pre,slen
    pre=[0]*29           #清空
    v = ' '+input()
    slen = len(v)-1
    for i in range(1,slen+1):		 
        if v[i]>='A' and v[i]<='Z': s[i] = -ord(v[i])   #字符不能直接相减，需要用ord()
        else:
            if pre[ord(v[i])-ord('a')+1]>0: s[i]=i-pre[ord(v[i])-ord('a')+1]
            else: s[i] = 0
            pre[ord(v[i])-ord('a')+1]=i
def initp():
    global pre,plen
    pre=[0]*29
    v=' '+input()
    plen=len(v)-1
    for i in range(1,plen+1):	
        if v[i]>='A' and v[i]<='Z': p[i]=-ord(v[i])
        else:
            if pre[ord(v[i])-ord('a')+1]>0:  p[i]=i-pre[ord(v[i])-ord('a')+1]
            else: p[i]=0
            pre[ord(v[i])-ord('a')+1]=i
def getNext():                       
    for i in range(1,plen):
        j = Next[i];                  
        while j>0 and check(p[i+1],p[j+1],j)==0:   j = Next[j]              
        if check(p[i+1],p[j+1],j)>0:   Next[i+1] = j+1
        else:             Next[i+1] = 0        
def kmp():
    ans = 0
    j = 0
    for i in range(0,slen):                                    
        while j>0 and check(s[i+1],p[j+1],j)==0:  j=Next[j] 
        if check(s[i+1],p[j+1],j)>0:   j+=1        
        if j == plen:  
            ans+=1
            j=Next[j]
    print(ans)                       
Q=int(input())
for i in range(Q):  inits();  initp();  getNext();  kmp()

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第9章编程作业.cpp
//编程作业 lanqiaoOJ 1433

#include<bits/stdc++.h>
using namespace std;
const int N=1e6+9;
char v[N];     // 读入字符串
int slen,plen,s[N],p[N],pre[29],Next[N];
bool check(int a,int b,int j) {
	if(a<0||b<0) return a==b;			//大写字母
	return a==b||(b==0&&a>j);			//小写字母
}
void inits(){  //转化母串s
    scanf("%s",v+1);
    slen = strlen(v+1);
	memset(pre,0,sizeof(pre));
	for(int i=1;i<=slen;i++)				//预处理s串
		if(v[i]>='A'&&v[i]<='Z')  s[i]=-v[i];  //大写字母记为负数
        else{
			if(pre[v[i]-'a'+1]>0) s[i]=i-pre[v[i]-'a'+1];
            else s[i]=0;
            pre[v[i]-'a'+1]=i;
        }
}
void initp(){                     //转化模式串P
	scanf("%s",v+1);  plen=strlen(v+1);
	memset(pre,0,sizeof(pre));
	for(int i=1;i<=plen;i++)				//预处理t串
		if(v[i]>='A'&&v[i]<='Z') p[i]=-v[i];
		else{
            if(pre[v[i]-'a'+1]>0)   p[i]= i-pre[v[i]-'a'+1];
            else p[i]=0;
            pre[v[i]-'a'+1]=i;
		}
}
void getNext(){    //Next[]模板
	Next[0]=Next[1]=0;
	for(int i=1;i<plen;i++) {
	    int j = Next[i];
		while(j&&!check(p[i+1],p[j+1],j))
            j=Next[j];
		if(check(p[i+1],p[j+1],j))
		    Next[i+1] = j+1;
		else  Next[i+1] = 0;
	}
}
void kmp() {      //kmp模板
	int ans=0;
	int j=0;
	for(int i=0;i<slen;i++) {
		while(j&&!check(s[i+1],p[j+1],j))
            j=Next[j];
		if(check(s[i+1],p[j+1],j))
            j++;
		if(j==plen)
            ans++,j=Next[j];  //找到一个匹配，继续找下一个
	}
	cout<<ans<<endl;
}
int main() {
	int Q; cin>>Q;
	while(Q--){ inits();  initp();  getNext();  kmp();}
	return 0;
}
